components/resource_metering/src/model.rs

```diff
@@ -64,10 +65,36 @@ impl Default for RawRecords {
             begin_unix_time_secs: now_unix_time.as_secs(),
             duration: Duration::default(),
             records: HashMap::default(),
+            others: RawRecord::default(),
         }
     }
 }
 
+impl RawRecords {
+    /// Keep a maximum of `k` self.records and aggregate the others into self.others.
+    pub fn keep_top_k(&mut self, k: usize) {
+        if self.records.len() <= k {
+            return;
+        }
+        let mut buf = STATIC_BUF.with(|b| b.take());
+        buf.clear();
+        // Find kth top cpu time.
+        for record in self.records.values() {
+            buf.push(record.cpu_time);
+        }
+        pdqselect::select_by(&mut buf, k, |a, b| b.cmp(a));
+        let kth = buf[k];
+        // Evict records with cpu time less or equal than `kth`
+        let others = &mut self.others;
+        let evicted_records = self.records.drain_filter(|_, r| r.cpu_time <= kth);
+        // Record evicted into others
+        for (_, record) in evicted_records {
+            others.merge(&record);
+        }
+        STATIC_BUF.with(move |b| b.set(buf));
+    }
+}
+
 /// Resource statistics.
 ///
 /// TODO(mornyx): Optimize to Vec<Item{timestamp,cpu_time,read_keys,write_keys}>
```

Comment:
```suggestion
```

Reply:
We still have:
```rust
pub struct Record {
    pub timestamps: Vec<u64>,
    pub cpu_time_list: Vec<u32>,
    pub read_keys_list: Vec<u32>,
    pub write_keys_list: Vec<u32>,
    pub total_cpu_time: u32,
}
```

This TODO refers to here.